### ISSUE TITLE
gotos in functions with variable SP are extra harmful

### ISSUE THREAD

**Issue by laubzega** [cc65/cc65#756]
Let's start with a test case:
```C
#include <stdio.h>

int main(void)
{
        if (2 == 2)
        {
                int var1;
                var1 = 0xf00d;
                printf("var1: %x\n", var1);
                goto skip;
        back:
                printf("var1: %x\n", var1);
        }
        else
        {
                int var2;
        skip:
                var2 = 0xdead;
                printf("var2: %x\n", var2);
                goto back;
        }
}
```
Expected output (gcc):
```
var1: f00d
var2: dead
var1: f00d
```
Actual:
```
var1: f00d
var2: dead
var1: dead
```
The reason for the observed behavior is that the value of SP changes during function execution (due to automatic variables being allocated and deallocated), yet is assumed to be constant when jumping around. If you thus goto between code areas where SP points to different locations (or same locations, but corresponding to a different variable - as in the test case) the result are trashed variables or spectacular crashes, especially if SP is not restored to its correct value ahead of function's finalization.

Additionally, the warning about unreachable code generated by the compiler is incorrect (perhaps another issue should be created about it).

**Comment by polluks** [comment]
It's a well-known bug... https://www.cc65.org/doc/BUGS

**Comment by oliverschmidt** [comment]
@laubzega: Thanks for the detailed bug report.
@polluks: Thanks for the hint/link.

I think it makes sense to keep this open - although it's the opposite of new ;-)

**Comment by groessler** [comment]
Funnyly, I had a problem with "gotos" yesterday, where I scratched my head. 

Look at the files in the attached zip. The program display/sets the system time. When setting the time, it also displays the time after setting for verification.
The "bad" version uses a "goto" to the display-time "if" clause after setting the time. It will display wrong values, typically wrong day-of-month and wrong year.
The "good" version has the display part in a function and shows the correct just-set time.

AFAICT, there are no "scope variables" involved in the bad case.

You'll need SpartaDOS-X on Atari to run the programs.

[sdxtime.zip](https://github.com/cc65/cc65/files/2398646/sdxtime.zip)

**Comment by laubzega** [comment]
While the testcase still behaves the same, it does not violate C89 (3.1.2.4). Other problems with goto (related to jumping into blocks with a different SP) have been resolved by PR #759. Closing.
